src/wallet/rpcwallet.cpp

```diff
@@ -161,8 +161,8 @@ UniValue getnewaddress(const JSONRPCRequest& request)
 
     // Generate a new key that is added to wallet
     CPubKey newKey;
-    if (!pwallet->GetKeyFromPool(newKey)) {
-        throw JSONRPCError(RPC_WALLET_KEYPOOL_RAN_OUT, "Error: Keypool ran out, please call keypoolrefill first");
+    if (!pwallet->GetKeyFromPool(newKey) || (pwallet->IsHDEnabled() && !pwallet->HasUnusedKeys(GetArg("-keypoolcritical", DEFAULT_KEYPOOL_CRITICAL)))) {
```

Comment:
In commit "[wallet] return error from getnewaddress if keypool drops to critical"

Maybe compare against -keypoolmin instead of -keypoolcritical to be more conservative, and because otherwise this change will have no effect (because the node will be shut down) unless `-bypasskeypoolcritical` happens to be in use. Also I think it's good just for sake of simplicity if `-bypasskeypoolcritical` and `-keypoolcritical=0` are synonyms for the same thing. Here you are introducing a subtle difference.

Either way, `-keypoolmin` or `-keypoolcritical` documentation could be updated to mention this new behavior, too.